### ISSUE TITLE
Document the relation between fractions, swarm modules and wildfly stock subsystems

### ISSUE THREAD

**Issue by heiko-braun** [wildfly-swarm/wildfly-swarm#95]
Fractions seem to provide custom modules that mix in fraction specific API's and most often refer to stock wildfly modules. I.e. the IO fraction:

```
<module xmlns="urn:jboss:module:1.3" name="org.wildfly.swarm.runtime.io">
  <resources>
    <artifact name="org.wildfly.swarm:wildfly-swarm-runtime-io:${project.version}"/>
  </resources>

  <dependencies>
    <module name="org.wildfly.swarm.io"/>
    <module name="org.wildfly.swarm.container"/>
    <module name="org.wildfly.swarm.runtime.container"/>

    <module name="org.wildfly.extension.io"/>
  </dependencies>
</module>

```

Can someone briefly explain the design rational behind this approach? I.e why are dedicated swarm modules needed? What do the swarm modules provide beyond the stock WF modules? Does this result a set modules bootstrapped by swarm which contains an embedded Wildfly (that uses it's own set of modules?)

**Comment by emag** [comment]
+1 and related #44

**Comment by bobmcwhirter** [comment]
I'll try to reply to #96 alongside this.

For more fractions, there are 2 pieces.  foo.jar and runtime-foo.jar.

The foo.jar (from api/foo/ in the build) is responsible for providing

* Any required module.xml from wf-swarm or WildFly itself
* The APIs (possibly to be generated in the future) to be used from the user-provided main() to configure the container (doing the stuff you would normally do in standalone.xml)

When a user adds `<artifactId>wildfly-swarm-foo</artifactId>` to his project's pom.xml, it gets added to the classpath, and our classpath-centric ModuleFinder stuff can then find both -swarm and regular wildfly module.xmls.

This replaces the static $JBOSS_HOME/modules/... with the superset of all modules brought in from all `wildfly-swarm-*` jars on the classpath.  

So, when the user adds wildfly-swarm-jaxrs, they also get wildfly-swarm-undertow, which also grabs wildfly-swarm-io, etc, and so the "complete" tree of module.xml is available, on the classpath.

When the user executes his `main()` directly from his IDE or from `mvn wildfly-swarm:run`, having all of that on the classpath is important, since we have no $JBOSS_HOME, and he's running his main() directly, not through any specialized runner.

Within his `main()`, which is executing from the context of the normal, default, system AppClassLoader, he constructs a new `Container`.

The `Container` is our first real entry-point into the JBoss-Modules system, and begins the process of working with the corresponding `wildfly-swarm-runtime-*` jars, which do *not* live in the user's classpath, but rather on the "inside" of JBoss-Modules loading.

When the user starts the `container`, he's really working with a facade on the "outside", which starts the `RuntimeServer` on the "inside", which can use all of the wildfly modules to do classloading, including access to baseline modules such as `org.jboss.as.server` which contains the real entry-point into embedding WildFly inside the user's `main()`.

Additionally, within the user's `main()`, he can use the various configuration APIs to setup subsystems (fractions), such as Logging, Clustering, etc.  These APIs, since they live "outside" of the jboss-modules stuff, directly accessible from the user's `main()` executing within the context of the system AppClassLoader, they cannot access stuff like the DMR `ModelNode` bits.  Instead, they are simply opaque configuration object trees that ultimately boil down to collections and scalar values.

Within `RuntimeServer` is some machinations involving service-loaders in order to bring in the `FooConfiguration` class on the "inside" of the classloader wall, which can consume the FooFraction config tree, and convert it into the relevant `List<ModelNode>` that matches what would happen in standalone.xml.

All of that being said, it really only address the aspects of running the user's `main()` from either an IDE or `mvn wildfly-swarm:run`.

When the user constructs an uber-jar, (myapp-swarm.jar) and executes it via `java -jar myapp-swarm.jar`, the uber-jar does *not* have all of the wildfly-swarm-foo.jar on the classpath.  The only thing in the system AppClassLoader's classpath is `myapp-swarm.jar`.

In that case, we actually setup the `MANIFEST.MF` to run our special bootstrapping `main()` first to figure out the user's `main()`, and work on adding the same things that would've been on the original classpath to the classloader that runs the user-provided `main()` within the fat-jar.

This is why the public/outside module descriptors look like:

    <module xmlns="urn:jboss:module:1.3" name="org.wildfly.swarm.clustering">

      <dependencies>
        
        <system export="true">
          <paths>
            <path name="org/wildfly/swarm/clustering"/>
          </paths>
        </system>

        
        <module name="org.wildfly.swarm.bootstrap" optional="true" export="true">
          <exports>
            <include path="org/wildfly/swarm/clustering"/>
          </exports>
        </module>

        <module name="org.wildfly.swarm.clustering" slot="runtime"/>

      </dependencies>
    </module>

In the IDE/mvn case, the `org.wildfly.swarm.clustering` classes are on the system classloader, so the `module.xml` just exposes them as the jboss-module named `org.wildfly.swarm.clustering:main`.

When bootstrapped through a fat-jar, we've actually added the contents of `wildfly-swarm-clustering.jar` to the `org.wildfly.swarm.bootstrap:main` module, and this `module.xml` just re-exports them.

This allows us to refer to the `org.wildfly.swarm.clustering.**` classes through the same module name, regardless of how we got booted up.

The reason some of our module.xml seem to just reference normal wildfly modules is that we have the `wildfly-swarm-fraction-plugin`, which process each of our `api/*/pom.xml projects`.  It walks the `module.xml` it finds in the project, and then transitively attempts to pull in other `module.xml` it needs to successfully have a full modules directory.  The transitive solution may use either other wildfly-swarm-foo.jar's, if they are dependencies, or it will slurp them from wildfly-*-feature-pack.zip's that are mentioned as dependencies.

The goal is that the collection of wildfly-swarm-foo.jar's referenced by a user's project has all the necessary and sufficient module.xml's required by the embedded WildFly.  If we didn't list out the WildFly modules that we require, they would not be pulled in through the transitive solution, and when we attempting to boot, it would complain that `org.wildfly.clustering.api:main` was missing, or whatnot.

Bottom line:

In wildfly-swarm, the projects under `api/*` can be considered classpath artifacts, available within the user's main(), but possibly being used outside of the jboss-modules classloader magic.

The projects under `runtime/*` are *always* referenced within the jboss-modules classloader magic.

And a collection of ServiceLoader descriptors tie them all together, and acts as automation to imply the equivalent of `<extension...>` from standalone.xml and to convert the "outside" configuration objects into "inside" `ModelNode` objects.

**Comment by bobmcwhirter** [comment]
All this being said, this has evolved organically and may not actually be a great/wonderful/good solution. When we first started, we solved the run-inside-an-uberjar use-case, but completely neglected the *run my main() from the IDE* case.  So a lot of this was to support that also.  The JBDS guys pushed back on requiring a maven build before execution, since it was indeed pretty crappy.

Now the right-click-and-run of a user-provided `main()` works from the IDE.  

The could certainly be opportunities to un-complicate this.  But jboss-modules and classloading in general may prevent much un-complicationing.

**Comment by bobmcwhirter** [comment]
I think this is being covered by the documentation at http://wildfly-swarm.gitbooks.io/
